File: app/src/main/java/com/example/android/securenote/crypto/RSAHardwareEncryptor.java
package com.example.android.securenote.crypto;

import android.content.Context;
import android.content.SharedPreferences;
import android.security.KeyChain;
import android.security.KeyPairGeneratorSpec;
import android.util.Base64;
import android.util.Base64InputStream;
import android.util.Base64OutputStream;
import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import java.security.InvalidAlgorithmParameterException;
import java.security.Key;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.UnrecoverableEntryException;
import java.security.cert.CertificateException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;
import java.util.Calendar;
import java.util.Date;

import javax.crypto.Cipher;
import javax.crypto.CipherInputStream;
import javax.crypto.CipherOutputStream;
import javax.security.auth.x500.X500Principal;

public class RSAHardwareEncryptor {
    private static final String TAG =
            RSAHardwareEncryptor.class.getSimpleName();
    private static final String PROVIDER_NAME = "AndroidKeyStore";
    private static final String KEY_ALGORITHM = "RSA";
    private static final String ENCRYPTION_ALGORITHM = "RSA/ECB/PKCS1Padding";

    private static final String KEY_PUBLIC = "publickey";
    private static final String KEY_ALIAS = "secureKeyAlias";

    private SharedPreferences mPublicKeyStore;

    public RSAHardwareEncryptor(Context context) {
        mPublicKeyStore = context.getSharedPreferences(
                "publickey.store", Context.MODE_PRIVATE);
        try {
            if (!mPublicKeyStore.contains(KEY_PUBLIC)) {
                generatePrivateKey(context);
                Log.d(TAG, "Generated hardware-bound key");
            } else {
                Log.d(TAG, "Hardware key pair already exists");
            }
        } catch (Exception e) {
            throw new RuntimeException("Unable to generate key material.");
        }
    }

    /**
     * Return a cipher text blob of encrypted data, Base64 encoded.
     *
     * @throws GeneralSecurityException
     * @throws IOException
     */
    public void encryptData(byte[] data, OutputStream out) throws
            GeneralSecurityException, IOException {
        Key key = retrievePublicKey();
        Cipher cipher = Cipher.getInstance(ENCRYPTION_ALGORITHM);
        cipher.init(Cipher.ENCRYPT_MODE, key);

        //Encode output to file
        out = new Base64OutputStream(out, Base64.NO_WRAP);
        //Encrypt output to encoder
        out = new CipherOutputStream(out, cipher);

        try {
            out.write(data);
            out.flush();
        } finally {
            out.close();
        }
    }

    /**
     * Return decrypted data from the received cipher text blob.
     *
     * @throws GeneralSecurityException
     * @throws IOException
     */
    public byte[] decryptData(InputStream in) throws
            GeneralSecurityException, IOException {
        Key key = retrievePrivateKey();
        Cipher cipher = Cipher.getInstance(ENCRYPTION_ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, key);

        //Decode input from file
        in = new Base64InputStream(in, Base64.NO_WRAP);
        //Decrypt input from decoder
        in = new CipherInputStream(in, cipher);

        return readFile(in).getBytes();
    }

    public Key retrievePublicKey() throws
            NoSuchAlgorithmException, InvalidKeySpecException {
        String encodedKey = mPublicKeyStore.getString(KEY_PUBLIC, null);
        if (encodedKey == null) {
            throw new RuntimeException("Expected valid public key!");
        }

        byte[] publicKey = Base64.decode(encodedKey, Base64.NO_WRAP);
        return KeyFactory.getInstance(KEY_ALGORITHM)
                        .generatePublic(new X509EncodedKeySpec(publicKey));
    }

    public Key retrievePrivateKey() throws
            GeneralSecurityException, IOException {
        KeyStore ks = KeyStore.getInstance(PROVIDER_NAME);
        ks.load(null);
        KeyStore.Entry entry = ks.getEntry(KEY_ALIAS, null);
        if (!(entry instanceof KeyStore.PrivateKeyEntry)) {
            Log.w(TAG, "Not an instance of a PrivateKeyEntry");
            return null;
        }

        return ((KeyStore.PrivateKeyEntry) entry).getPrivateKey();
    }

    private void generatePrivateKey(Context context) throws
            GeneralSecurityException {
        Calendar cal = Calendar.getInstance();
        Date now = cal.getTime();
        cal.add(Calendar.YEAR, 1);
        Date end = cal.getTime();

        KeyPairGenerator kpg = KeyPairGenerator.getInstance(KEY_ALGORITHM, PROVIDER_NAME);
        kpg.initialize(new KeyPairGeneratorSpec.Builder(context)
                .setAlias(KEY_ALIAS)
                .setStartDate(now)
                .setEndDate(end)
                .setSerialNumber(BigInteger.valueOf(1))
                .setSubject(new X500Principal("CN=" + KEY_ALIAS))
                .build());

        //Generate and bind the private key to hardware
        KeyPair kp = kpg.generateKeyPair();

        //Persist the public key
        PublicKey publicKey = kp.getPublic();
        String encodedKey = Base64.encodeToString(publicKey.getEncoded(), Base64.NO_WRAP);
        mPublicKeyStore.edit().putString(KEY_PUBLIC, encodedKey).apply();
    }

    private String readFile(InputStream in) throws IOException {
        InputStreamReader reader = new InputStreamReader(in);
        StringBuilder sb = new StringBuilder();

        char[] inputBuffer = new char[2048];
        int read;
        while ((read = reader.read(inputBuffer)) != -1) {
            sb.append(inputBuffer, 0, read);
        }

        return sb.toString();
    }
}

File: app/src/main/java/com/example/android/securenote/crypto/RSAHardwareEncryptor.kt
package com.example.android.securenote.crypto

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.math.BigInteger
import java.security.*
import java.security.spec.InvalidKeySpecException
import java.security.spec.RSAKeyGenParameterSpec
import java.util.*
import javax.security.auth.x500.X500Principal


class RSAHardwareEncryptor(context: Context) {
    var masterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()

    //Persistent location where we will save the public key
    private val sharedPreferences: SharedPreferences = EncryptedSharedPreferences.create(
        context,
        "secret_shared_prefs",
        masterKey,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    init {
        try {
            if (!sharedPreferences.contains(KEY_PUBLIC)) {
                generatePrivateKey()
                Log.d(TAG, "Generated hardware-bound key")
            } else {
                Log.d(TAG, "Hardware key pair already exists")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Unable to generate key material.", e)
            throw RuntimeException("Unable to generate key material.")
        }
    }

    /**
     * Create a self-signed certificate and private key in hardware storage.
     * Persist the (non-secret) public key into SharedPreferences.
     *
     * @throws GeneralSecurityException
     */
    @Throws(GeneralSecurityException::class)
    private fun generatePrivateKey() {
        val keyPairGenerator =
            KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, PROVIDER_NAME)
        keyPairGenerator.initialize(
            KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setAlgorithmParameterSpec(RSAKeyGenParameterSpec(2048, RSAKeyGenParameterSpec.F4))
                .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_PKCS1)
                .setDigests(
                    KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA384,
                    KeyProperties.DIGEST_SHA512
                )
                //.setUserAuthenticationRequired(true) <-- requires fingerprint
                .build()
        )
        val keyPair = keyPairGenerator.generateKeyPair()

        // Persist the public key
        val publicKey = keyPair.public
        val encodedKey = Base64.encodeToString(publicKey.encoded, Base64.NO_WRAP)
        sharedPreferences.edit().putString(KEY_PUBLIC, encodedKey).apply()
    }

    /**
     * Return a cipher text blob of encrypted data, Base64 encoded.
     *
     * @throws GeneralSecurityException
     * @throws IOException
     */
    @Throws(GeneralSecurityException::class, IOException::class)
    fun encryptData(data: ByteArray, out: OutputStream) {
        /*
         * TODO: Encryption Lab:
         * Obtain the public key for encryption
         * Create and init a Cipher (with ENCRYPTION_ALGORITHM)
         * Wrap the supplied stream in another provides Base64 encoding
         * Wrap the encoding stream in another that encrypts with cipher
         * Write the supplied data to the streams.
         */
    }

    /**
     * Return decrypted data from the received cipher text blob.
     *
     * @throws GeneralSecurityException
     * @throws IOException
     */
    @Throws(GeneralSecurityException::class, IOException::class)
    fun decryptData(input: InputStream): ByteArray? {
        /*
         * TODO: Encryption Lab:
         * Obtain the private key for decryption
         * Create and init a Cipher (with ENCRYPTION_ALGORITHM)
         * Wrap the supplied stream in another parses Base64 encoding
         * Wrap the encoding stream in another that decrypts with cipher
         * Read the stream fully and return the decrypted bytes
         */
        return null
    }

    @Throws(NoSuchAlgorithmException::class, InvalidKeySpecException::class)
    fun retrievePublicKey(): Key? {
        /*
         * TODO: Encryption Lab:
         * Get the encoded key from SharedPreferences
         * Decode the key (from Base64) to raw bytes
         * Return a public key instance from the bytes using KeyFactory
         */
        return null
    }

    fun retrievePrivateKey(): PrivateKey? {
        /*
         * TODO: Encryption Lab:
         * Obtain an instance of AndroidKeyStore and load it
         * Get the private key alias and return the key
         */
        return null
    }

    /* Helper method to parse a file stream into memory */
    @Throws(IOException::class)
    private fun readFile(input: InputStream): String {
        val reader = InputStreamReader(input)
        val sb = StringBuilder()

        val inputBuffer = CharArray(2048)
        var read: Int = reader.read(inputBuffer)
        while (read != -1) {
            sb.append(inputBuffer, 0, read)
            read = reader.read(inputBuffer)
        }

        return sb.toString()
    }

    companion object {
        private val TAG = RSAHardwareEncryptor::class.java.simpleName
        private const val PROVIDER_NAME = "AndroidKeyStore"
        private const val KEY_ALGORITHM = "RSA"
        private const val ENCRYPTION_ALGORITHM = "RSA/ECB/PKCS1Padding"

        //Preferences alias for the public key
        private const val KEY_PUBLIC = "publickey"

        //KeyStore alias for the private key
        private const val KEY_ALIAS = "secureKeyAlias"
    }
}
